### PR TITLE
Provide a general way to write an StreamMessage to a file

### PR DESCRIPTION
**Pull request by wooseongshin**
#### Provide a general way to write an StreamMessage to a file

@trustin suggested an awesome idea to write a StreamMessage<T> into a File or Path. 

```java
interface StreamMessage<T> {
    default CompletableFuture<Void> writeTo(Path destination, Function<? super T, ? extends HttpData> mapper, options...) {
        // Delegate to StreamMessages.writeTo(...)
    }
}
```

### Modifications:

- Add `StreamMessage.writeTo() `
- Edit the StreamMessageTest by referring to the existing streamMessagesTest.

### Result:

- We can now write file using `StreamMessge.writeTo()`

### CODE REVIEW

File: core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java [hunk 1]
```diff
@@ -269,6 +276,30 @@ public void onComplete() {
         await().untilAsserted(() -> assertThat(buf.refCnt()).isZero());
     }
 
+    @TempDir
+    static Path tempDir;
```
Comment (injae-kim) [inline core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java]: Can we remove this or just move inside of `test`?
(If this static variable is necessary, then we should locate it on top of this class same with other variables)

File: core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java [hunk 2]
```diff
@@ -269,6 +276,30 @@ public void onComplete() {
...
+
+    @Test
+    void defaultWriteStream() throws IOException {
```
Comment (injae-kim) [inline core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java]:
```suggestion
    void writeToFile() throws IOException {
```

File: core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java [hunk 3]
```diff
@@ -269,6 +276,30 @@ public void onComplete() {
...
+        final List<ByteBuf> bufs = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            bufs.add(Unpooled.wrappedBuffer(Integer.toString(i).getBytes()));
+        }
+        final HttpData[] httpData = bufs.stream().map(HttpData::wrap).toArray(HttpData[]::new);
+        final byte[] expected = Arrays.stream(httpData)
+                .map(HttpData::array)
+                .reduce(Bytes::concat).get();
```
Comment (injae-kim) [inline core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java]:
```suggestion
        final HttpData[] httpData = bufs.stream().map(HttpData::wrap).toArray(HttpData[]::new);
        final byte[] expected = Arrays.stream(httpData)
                .map(HttpData::array)
                .reduce(Bytes::concat).get();
                
```
nit. need space

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java [hunk 4]
```diff
@@ -676,4 +677,24 @@ default EventExecutor defaultSubscriberExecutor() {
...
+     *  ByteBuf[] bufs = new ByteBuf[10];
+     *  for(int i=0; i<10; i++){
+     *             bufs[i] = Unpooled.wrappedBuffer(Integer.toString(i).getBytes());
+     *   }
+     *  StreamMessage<ByteBuf> streamMessage = StreamMessage.of(bufs);
+     *  streamMessage.writeTo( x -> HttpData.wrap())
+     *  }</pre></p>
+     */
+    default CompletableFuture<Void> writeTo(
+            Function<? super T, ? extends HttpData> mapper, Path destination, OpenOption... options) {
```
Comment (injae-kim) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java]:
```suggestion
    default CompletableFuture<Void> writeTo(Function<? super T, ? extends HttpData> mapper, 
                                            Path destination, OpenOption... options) {
```
nit. use same indent with `StreamMessages#writeTo`

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java [hunk 5]
```diff
@@ -676,4 +677,24 @@ default EventExecutor defaultSubscriberExecutor() {
...
+    /**
+     *  Provide general way to use {@link StreamMessages} by applying the specified {@link Function}.
+     *
+     *  <p>Example:<pre>{@code
+     *  ByteBuf[] bufs = new ByteBuf[10];
+     *  for(int i=0; i<10; i++){
+     *             bufs[i] = Unpooled.wrappedBuffer(Integer.toString(i).getBytes());
+     *   }
+     *  StreamMessage<ByteBuf> streamMessage = StreamMessage.of(bufs);
+     *  streamMessage.writeTo( x -> HttpData.wrap())
```
Comment (injae-kim) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java]: *  ByteBuf[] bufs = new ByteBuf[10]; // -> List<ByteBuf> bufs = new ArrayList<>();
     *  for(int i=0; i<10; i++){ // -> for (int i = 0; i < 10; i++)
     *       bufs.add(Unpooled.wrappedBuffer(Integer.toString(i).getBytes())); // indent, 4 space for tab
     *   }
     *  StreamMessage<ByteBuf> streamMessage = StreamMessage.of(bufs);
     *  streamMessage.writeTo( x -> HttpData.wrap()) // (x -> HttpData.wrap(x))
     * + assert~ // where is assert?

File: core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java (hunk 2 above)
Comment (injae-kim) [reply core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java]: nit. test naming

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java [hunk 6]
```diff
@@ -676,4 +677,24 @@ default EventExecutor defaultSubscriberExecutor() {
         requireNonNull(function, "function");
         return new RecoverableStreamMessage<>(this, function, /* allowResuming */ true);
     }
+
+    /**
+     *  Provide general way to use {@link StreamMessages} by applying the specified {@link Function}.
```
Comment (injae-kim) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java]:
```suggestion
     * Provide general way to use {@link StreamMessages} by applying the specified {@link Function}.
```
nit. should use 1 space here

Comment (injae-kim) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java]: /** 
  * Writes this {@link StreamMessage} to the given {@link Path} with {@link OpenOption}s.
  * See {@link StreamMessages#writeTo} for the details.

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java (hunk 4 above)
Comment (wooseongshin) [reply core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java]: > nit. use same indent with `StreamMessages#writeTo`

ok

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java (hunk 5 above)
Comment (wooseongshin) [reply core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java]: test

**Comment by wooseongshin** [comment]
test

**Pull request closed without merging**